### PR TITLE
Rust activation

### PR DESCRIPTION
**Pull request by mingwandroid**
I've used these on AD for a while, but we don't have a great many rust packages. I test every so often against `alacritty` and `ripgrep`, so these could do with more eyes and thoughts (and tests and ..?)

The idea is the same as other compiler activation packages. We make cargo config files that point rustc at our compilers basically.

I should update to the latest rust first and for now, win-32 is supported (I'll exclude that for conda-forge), still I want to get early input if possible hence putting this up.

### CODE REVIEW

File: recipes/rust-activation/activate.bat [hunk 1]
```diff
@@ -0,0 +1,9 @@
+set CARGO_HOME=%CONDA_PREFIX%.cargo.win
+set CARGO_CONFIG=%CARGO_HOME%\config
+set RUSTUP_HOME=%CARGO_HOME%\rustup
+set PATH=%CARGO_HOME%\bin:%PATH%
+
+if not exist "%CARGO_HOME%" mkdir "%CARGO_HOME%"
+
+echo [target.x86_64-pc-windows-msvc]>> %CARGO_CONFIG%
```
Comment (msarahan) [inline recipes/rust-activation/activate.bat]: This hardcoded value may conflict with a non default cxx_compiler. I may be preferable to require an explicit rust toolchain entry in cbc.yaml that is zipped with the compiler settings.

It's probably hard enough to get this right and enough of a corner case that you should ignore it for now.

**Comment by mingwandroid** [comment]
I need to push some fixes for windows. Don't merge for now. Cheers.

That string is enforced by the rust compiler being used. You can actually add multiple ones if you have cross compilers available!

File: recipes/rust-activation/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      files:  # [win]
+        - forge_test.sh  # [win]
+      commands:  # [win]
+        - rustc --help  # [win]
+        - cargo --help  # [win]
+
+about:
+  home: https://www.rust-lang.org
+  summary: A safe systems programming language (conda activation scripts)
+  license: BSD
```
Comment (chrisburr) [inline recipes/rust-activation/meta.yaml]:
```suggestion
  license: BSD-3-Clause
```

File: recipes/rust-activation/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,56 @@
+{% set version = "1.40.0" %}
+
+package:
+  name: rust_{{ target_platform }}-suite
+  version: {{ version }}
+
+build:
+  number: 1
```
Comment (chrisburr) [inline recipes/rust-activation/meta.yaml]: (To save anyone else from checking the same thing.) This needs to start at `1` to match with what is already in `defaults`.

File: recipes/rust-activation/forge_test.sh [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+rustc --help
+rustdoc --help
+cargo --help
+
+echo "#!/usr/bin/env bash"                         > ./cc
+if [[ ${target_platform} =~ linux.* ]]; then
+  echo "x86_64-conda_cos6-linux-gnu-cc \"\$@\""   >> ./cc
+elif [[ ${target_platform} == osx-64 ]]; then
+  echo "x86_64-apple-darwin13.4.0-clang \"\$@\""  >> ./cc
+  export CONDA_BUILD_SYSROOT=/opt/MacOSX10.10.sdk
```
Comment (chrisburr) [inline recipes/rust-activation/forge_test.sh]: I guess this means rust needs 10.10 or later? If so, see: https://conda-forge.org/docs/maintainer/knowledge_base.html#requiring-newer-macos-sdks

File: recipes/rust-activation/forge_test.sh [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+unset REQUESTS_CA_BUNDLE
+unset SSL_CERT_FILE
+
+rustc --help
+rustdoc --help
+cargo --help
+
+echo "#!/usr/bin/env bash"                         > ./cc
+if [[ ${target_platform} =~ linux.* ]]; then
```
Comment (chrisburr) [inline recipes/rust-activation/forge_test.sh]: I think the `if` can be replaced with this?
```suggestion
echo "$CC \"\$@\"" >> ./cc
```

File: recipes/rust-activation/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        - cargo --help
+        - time bash ./forge_test.sh  # [not win]
+  - name: rust-gnu_{{ target_platform }}  # [win]
+    script: bld.bat  # [win]
+    requirements:  # [win]
+      run:  # [win]
+        - rust-gnu  # [win]
+        - m2w64-toolchain  # [win]
+    test:  # [win]
+      files:  # [win]
```
Comment (chrisburr) [inline recipes/rust-activation/meta.yaml]: I think these files aren't used for this output?

File: recipes/rust-activation/forge_test.sh [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+  echo "x86_64-conda_cos6-linux-gnu-cc \"\$@\""   >> ./cc
+elif [[ ${target_platform} == osx-64 ]]; then
+  echo "x86_64-apple-darwin13.4.0-clang \"\$@\""  >> ./cc
+  export CONDA_BUILD_SYSROOT=/opt/MacOSX10.10.sdk
+fi
+cat cc
+chmod +x cc
```
Comment (chrisburr) [inline recipes/rust-activation/forge_test.sh]: Is this wrapper needed?

File: recipes/rust-activation/forge_test.sh (hunk 5 above)
Comment (mingwandroid) [reply recipes/rust-activation/forge_test.sh]: I believe `CC` may not be set at this point. It depends on the order of activation scripts being run for that to work!

File: recipes/rust-activation/forge_test.sh (hunk 4 above)
Comment (mingwandroid) [reply recipes/rust-activation/forge_test.sh]: Good point.

We should see if we can setup a virtual package for the macOS version and I should figure out what rust 1.40.0 requires as the minimum runtime. The rust compiler activation package should then impose a strong run_export on the virtual package. All very possible I think, now. I should do some digging.

**Comment by jakirkham** [comment]
cc @conda-forge/rust

File: recipes/rust-activation/forge_test.sh (hunk 7 above)
Comment (mingwandroid) [reply recipes/rust-activation/forge_test.sh]: Maybe not. It seemed to me that conda-build wasn't activating the test env correctly though but hopefully that was transient.

**Comment by jakirkham** [comment]
@chrisburr, would you have time to take another look? 🙂

**Comment by mingwandroid** [comment]
I think we should continue to work on this. Thoughts?

**Comment by jakirkham** [comment]
Yeah I removed the stale label already and asked Chris if he could take another look.

**Pull request merged**